### PR TITLE
Xorg libraries: new formula [WIP]

### PR DESCRIPTION
**Pull request by maxim-belkin**
Inspired by the pull request #891 and remembering my promise back in #826, I've finally decided to spent time to create `Xorg libraries` formula according to [LFS](http://www.linuxfromscratch.org/blfs/view/svn/x/x7lib.html). I've tested it on my machine, it works. The problem is that formula has to have a common URL that is used to deduce version number and Xorg libraries don't have a single url. I copied info from one of the Xorg "resources" (libFS I believe), so xorg is installed in a subfolder `1.0.7` instead of just `7`. So, this part needs fix and I'm not sure how to do that.

### CODE REVIEW

File: Library/Formula/xorg.rb [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+  end
+
+  ##### END of XPROTOCOL HEADERS 
+
+  resource "xcb-proto" do
+    url "http://xcb.freedesktop.org/dist/xcb-proto-1.11.tar.gz"
+    sha256 "d12152193bd71aabbdbb97b029717ae6d5d0477ab239614e3d6193cc0385d906"
+  end
+
+## for libxcb... not sure where it is needed there
```
Comment (rwhogg) [inline Library/Formula/xorg.rb]: Strictly speaking, libpthread-stubs isn't necessary on Linux. [BLFS](http://www.linuxfromscratch.org/blfs/view/svn/x/libxcb.html) actually recommends stripping the dependency on it out:

> sed "s/pthread-stubs//" -i configure: This sed removes dependency on libpthread-stubs package which is useless on Linux. 

I wasn't wholly comfortable with that, so I decided to just include it as a resource.

Comment (rwhogg) [inline Library/Formula/xorg.rb]: Ah, I see that you already accounted for that.

**Comment by rwhogg** [comment]
I'm getting the following error (ran this on a standalone-Linuxbrew Docker image)
> make[2]: Leaving directory '/tmp/xorg--libXpm20160306-30770-8m7m98/libXpm-3.5.11/src'
Making all in sxpm
make[2]: Entering directory '/tmp/xorg--libXpm20160306-30770-8m7m98/libXpm-3.5.11/sxpm'
  CC       sxpm.o
  GEN      sxpm.po
/bin/bash: xgettext: command not found
Makefile:651: recipe for target 'sxpm.po' failed
make[2]: *** [sxpm.po] Error 127
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory '/tmp/xorg--libXpm20160306-30770-8m7m98/libXpm-3.5.11/sxpm'
Makefile:467: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/tmp/xorg--libXpm20160306-30770-8m7m98/libXpm-3.5.11'
Makefile:375: recipe for target 'all' failed
make: *** [all] Error 2

Also, FYI https://github.com/Homebrew/homebrew/pull/49497 the xorg macros might be coming to plain Homebrew and it looks like they're discussing some of the same issues we are here.

File: Library/Formula/xorg.rb [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+  depends_on "doxygen"    =>  :optional if build.with?("doxygen") # to build API docs
+  depends_on "check"      =>  :optional if build.with?("check-libxcb") # to run tests
+  depends_on "libxslt"    =>  :optional # no explanation given
+
+
+# xcbproto: required dependencies
+  depends_on :python          # 
+  depends_on "libxml2"    =>  :optional if build.with?("check-xcbproto") # to run tests
+ #depends_on "libXdmcp"   => [:optional, :recommended] # built as a resource here
+# depends_on "pkg-config" => :build
```
Comment (rwhogg) [inline Library/Formula/xorg.rb]: May I ask why this is commented out?

File: Library/Formula/xorg.rb [hunk 3]
```diff
@@ -0,0 +1,470 @@
+class Xorg < Formula
+  desc "Xorg Libraries"
+  homepage "http://www.linuxfromscratch.org/blfs/view/svn/x/x7lib.html"
+  revision 7
```
Comment (rwhogg) [inline Library/Formula/xorg.rb]: I don't think we need this. Quoting from the [Formula Cookbook](https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/Formula-Cookbook.md#formula-cookbook):

> Occasionally, these updates require a forced-recompile of the formula itself or its dependents to either ensure formulae continue to function as expected or to close a security issue. This forced-recompile is known as a revision and inserted underneath the homepage/url/sha block.

File: Library/Formula/xorg.rb [hunk 4]
```diff
@@ -0,0 +1,470 @@
+class Xorg < Formula
+  desc "Xorg Libraries"
+  homepage "http://www.linuxfromscratch.org/blfs/view/svn/x/x7lib.html"
```
Comment (rwhogg) [inline Library/Formula/xorg.rb]: Strictly a matter of opinion, but I think http://www.x.org/wiki/ would make more sense for the homepage.

File: Library/Formula/xorg.rb (hunk 3 above)
Comment (maxim-belkin) [inline Library/Formula/xorg.rb]: Agreed. I was trying to make xorg install into `7` subdirectory.

File: Library/Formula/xorg.rb [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+  # tag "linuxbrew"
+
+  option "with-docs", "Build additional PDF or text documentation for the libXfont"
+  option "with-doxygen", "Generate API documentation (for libxcb)"
+  option "with-check-libxcb", "Run tests (for libxcb)"
+  option "with-check-xcbproto", "Run tests (for xcb-proto)"
+
+# Xorg: required dependencies
+  depends_on "fontconfig"
+# depends_on "libxcb"      # built as a resource here
```
Comment (rwhogg) [inline Library/Formula/xorg.rb]: Not sure exactly where to put it, but I think you're missing a dependency on `gettext`.

File: Library/Formula/xorg.rb (hunk 4 above)
Comment (maxim-belkin) [inline Library/Formula/xorg.rb]: Agree. I used this web page because I followed steps from there.

File: Library/Formula/xorg.rb (hunk 5 above)
Comment (maxim-belkin) [inline Library/Formula/xorg.rb]: could be. I will add it. I'll check where it comes from.

File: Library/Formula/xorg.rb (hunk 2 above)
Comment (maxim-belkin) [inline Library/Formula/xorg.rb]: Well, the main reason is that I was finishing it up at 5am.

**Comment by rwhogg** [comment]
@maxim-belkin I can confirm that, with the added dependency on `gettext`, this installs correctly on a standalone Linuxbrew Docker container.

Great work here!

File: Library/Formula/xorg.rb (hunk 3 above)
Comment (sjackman) [inline Library/Formula/xorg.rb]: Use
```ruby
version "7"
```

**Comment by sjackman** [comment]
Great work! Thanks Maxim and Bob. It's going to be very useful, I think.

What's the thoughts on one big formula with many resources vs one formula per library? I have a uninformed preference toward one formula per library and one `xorg` virtual package that depends on it, but I'd like to hear the benefits of both ways.

**Comment by maxim-belkin** [comment]
I'd prefer one virtual package and many individual ones as small packages are easy to update. But the trouble is that there are 60+ of them (of really small ones). We can probably create all the small ones and split them into meta-packages as in LFS and Xorg will use those meta-packages.

**Comment by rwhogg** [comment]
Given that so many of the individual formulae use the exact same `configure` flags and whatnot and aren't likely to be depended on directly, this single formula idea actually makes a lot of sense to me. (Ironically, me @maxim-belkin are both arguing for the opposite of what our respective PR's actually do!)

There's also the matter of boilerplate - all of these small formulae would probably duplicate details.

Finally, given that upstream Homebrew has PR's like https://github.com/Homebrew/homebrew/pull/49497, splitting up the formulae sounds like it could potentially conflict. By contrast, a single `xorg` formula probably won't do that, since XQuartz seems to be the de facto standard on Macs.

File: Library/Formula/xorg.rb [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+      system "make", "install"
+    end
+
+ 
+    res = %w[xtrans libX11 libXext libFS libICE libSM libXScrnSaver libXt libXmu libXpm libXaw libXfixes libXcomposite libXrender libXcursor libXdamage libfontenc libXfont libXft libXi libXinerama libXrandr libXres libXtst libXv libXvMC libXxf86dga libXxf86vm libdmx libpciaccess libxkbfile libxshmfence]
+
+    res.each do |r|
+      resource(r).stage do
+
+      case r
```
Comment (sjackman) [inline Library/Formula/xorg.rb]: A hash table could be used here rather than a `case` statement.

File: Library/Formula/xorg.rb [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+        system "./configure", "--prefix=#{prefix}", "--sysconfdir=#{prefix}/etc", "--localstatedir=#{prefix}/var", "--disable-static", "--disable-devel-docs"
+      when "libXt"
+        system "./configure", "--prefix=#{prefix}", "--sysconfdir=#{prefix}/etc", "--localstatedir=#{prefix}/var", "--disable-static", "--with-appdefaultdir=#{prefix}/X11/app-defaults"
+      else
+        system "./configure", "--prefix=#{prefix}", "--sysconfdir=#{prefix}/etc", "--localstatedir=#{prefix}/var", "--disable-static"
+      end
+
+        system "make"
+        system "make", "install"
+        system "ldconfig"
```
Comment (sjackman) [inline Library/Formula/xorg.rb]: `ldconfig` shouldn't be necessary. Linuxbrew uses `RPATH` rather than `ld.so.conf`.

**Comment by sjackman** [comment]
The boilerplate code could be deduplicated by using a `XFormula` base class. It would be easier to update the versions of the individual components if they were in separate formula. It could be a `Linuxbrew/x11` tap with the 60+ formula in it.

I am okay though with one big `xorg` package, or a perhaps a handful of packages as Maxim was suggesting. With one big formula, it might be easier to update it periodically on some sort of release schedule, and tie those updates to the release schedule of the XOrg project.

File: Library/Formula/xorg.rb (hunk 6 above)
Comment (maxim-belkin) [inline Library/Formula/xorg.rb]: I think `case` is easier to understand in this case because of `else`.

**Comment by maxim-belkin** [comment]
Interestingly, on OS X the formula uses `install -c -d` to create directories, while on Linux it uses `mkdir -p`. As a result, `xextproto` fails to build because of repeating directory creation somewhere...

**Comment by maxim-belkin** [comment]
Adding `depends_on "coreutils" => :build if OS.mac?` fixes it!

**Comment by sjackman** [comment]
Interesting, and how odd.

**Comment by sjackman** [comment]
And cool to hear that it builds on Mac OS! I was wondering whether it would.

**Comment by sjackman** [comment]
`version "7"`
Can you find out what versions of these libraries are used by `xorg 7.7`, use those same versions of the libraries, and change the version number to `7.7`?

**Comment by maxim-belkin** [comment]
I can certainly do that but... the packages in xorg 7.7 are outdated. The following statement can be found on [Xorg website](http://www.x.org/wiki/Releases/7.7/)

> Update: Many modules have had new releases since the X11R7.7 katamari - see the individual X.Org releases directory for downloads, and the xorg-announce archives for details on included changes. These updates may be included in a future 7.8 katamari release if one is ever assembled.

So, do we want to downgrade the packages? We could create a formula `xorg-7.7` and `xorg-latest`...

There is still a problem with Mac OS, actually. For whatever reason, `binutils` on a Mac do not provide `ld` (as well as `as` and `gprof`). Moreover, `gcc` package on a Mac does not create symlinks `gcc -> gcc-5`, `cpp -> cpp-5` that are created on a Linux. Absence of these cause all sorts of problems later on. I'm sure it is solvable though.

**Comment by maxim-belkin** [comment]
So, it looks like the trouble on OS X is caused by non-GNU ld. I will try to follow [these guides](https://github.com/cfenollosa/os-tutorial/tree/master/11-kernel-crosscompiler) to make it compile.

**Comment by maxim-belkin** [comment]
Question: I've compiled `binutils` on OS X so that they have ld and correct gcc. How do I use that `ld` instead of a system's one? For cc, I can use `--cc=gcc`, but what about `ld`? I've tried setting it inside of a formula (ENV["LD"] = "/path/to/ld") and it didn't work.

**Comment by sjackman** [comment]
Hi, Maxim. On Mac OS it should use the `binutils` provided by the host OS (aka XCode and Command Line Tools). If it can build with `clang`, great, and if not, add a `fails_with :clang` stanza. If it doesn't build with the host `binutils` I would forget about the Mac OS support for now and focus on Linux.

**Comment by sjackman** [comment]
If we're releasing one big X.Org formula, I want to avoid updating it super often and driving the users crazy with updates. If we're going to do only periodic releases, it would be better I think to tie our releases to X.Org releases. [X.Org 7.7](http://www.x.org/wiki/Releases/7.7/) was released on 2012-06-06, so it's pretty darn ancient. So, I agree I'd rather not use X.Org 7.7 What do you think we should do? What has Ubuntu done? How about we use the same versions as a particular Ubuntu release, such as trusty or wily.

**Comment by sjackman** [comment]
> Actually, problems on a Mac appear very close to the very end of compilation and this is where I felt that if we had small packages -- it would have been much faster to troubleshoot.

There's a number of things that may be easier with small packages. Packaging each library individually in `linxubrew/x11` tap is still appealing to me.

**Comment by maxim-belkin** [comment]
`linuxbrew/x11` seems like a good idea to me too. Let's do that. Will you create this tap first? I could then make a PR to it.  I'm not a fan of outdated packages so I vote for having the latest xorg.

**Comment by maxim-belkin** [comment]
There is a problem with Mac's `cpp`, not `clang`. We can indeed work on it after releasing a version for Linux.

**Comment by sjackman** [comment]
Homebrew already has a tap named `homebrew/x11`, which is for application that use X11 libraries.
To avoid confusion, I'm going to start a tap named `linuxbrew/xorg`, for the X.Org libraries themselves.

**Comment by sjackman** [comment]
Done: https://github.com/Linuxbrew/homebrew-xorg

**Comment by sjackman** [comment]
@rwhogg Which do you prefer, the monolithic `xorg` formula or the `Linuxbrew/xorg` tap?

**Comment by maxim-belkin** [comment]
Great, I've tapped it. I will start working on the formulas then.

**Comment by rwhogg** [comment]
I think a tap sounds good.

On March 8, 2016 11:07:33 AM PST, Shaun Jackman <notifications@github.com> wrote:
>@rwhogg Which do you prefer, the monolithic `xorg` formula or the
>`Linuxbrew/xorg` tap?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/Linuxbrew/linuxbrew/pull/892#issuecomment-193919900

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by sjackman** [comment]
Great.

**Comment by maxim-belkin** [comment]
Is there a clean way to define arguments shared across multiple formulas?
In all small formulas we will be using something like: 
```ruby
    args = %W[
      --prefix=#{HOMEBREW_PREFIX}/Cellar/xorg/latest
      --sysconfdir=#{HOMEBREW_PREFIX}/Cellar/xorg/latest/etc
      --localstatedir=#{HOMEBREW_PREFIX}/Cellar/xorg/latest/var
    ]
    system "./configure", *args
    system "make", "install"
```

Or do you think we should install formulas into their own folders (`--prefix=#{prefix}`)? LFS guide uses the same prefix for all the formulas. So, if we can't define common arguments in one place -- we have to make sure everything plays well.

**Comment by maxim-belkin** [comment]
I will use `#{prefix}` for now and then will change that if necessary.

**Comment by sjackman** [comment]
> Or do you think we should install formulas into their own folders (--prefix=#{prefix})?

Definitely put each formula in its own prefix.

**Comment by maxim-belkin** [comment]
Quick question: how to avoid "Error: Empty installation" in a meta-formula that does not have anything to do inside its ```def install```?

**Comment by maxim-belkin** [comment]
I've figured this out. I just need to create a file that is not recognized by Homebrew as unnesessary.

**Comment by maxim-belkin** [comment]
I'm almost done with the packages...

**Comment by sjackman** [comment]
> Quick question: how to avoid "Error: Empty installation" in a meta-formula that does not have anything to do inside its def install?

Add a `README.md` file that describes the purpose of the meta-formula. Put it either in `prefix` or `doc`.

> I'm almost done with the packages...
 Hide all checks

Great!

**Comment by maxim-belkin** [comment]
Looking at `Library/Homebrew/metafiles.rb` it looks like` README.md` will be ignored. So I just create an `html` file in there with different name.

**Comment by maxim-belkin** [comment]
Stylistic question:

With a single formula for `xorg` I used **`--sysconfdir=#{prefix}/etc`** and **`--localstatedir=#{prefix}/var`** which were ending up in 
**`[linuxbrew]/Cellar/xorg/latest/etc`**  and    **`[linuxbrew]/Cellar/xorg/latest/var`**.
With multiple formulas, I can imagine a situation when someone needs just a single formula not the entire set of xorg libraries. Therefore, a question: should I use
    * **`#{etc}`** and **`#{var}`** (i.e. *global* Homebrew folders),   or
    * **`#{prefix}/etc`** and **`#{prefix}/var`** (local to each formula),    or
    *  **`#{HOMEBREW_PREFIX}/Cellar/xorg/latest/etc`** and **`#{HOMEBREW_PREFIX}/Cellar/xorg/latest/var`** (*common* for all formulas)?

**Comment by sjackman** [comment]
I believe #{etc} and #{var} (i.e. global Homebrew folders).

**Comment by sjackman** [comment]
Moved to Linuxbrew/homebrew-xorg/issues/1.

**Comment by sjackman** [comment]
@tseemann Torsten, I recall you having some X issues with R. Have you tried Linuxbrew/xorg/xorg?

**Pull request closed without merging**